core/capabilities/compute/compute.go

```diff
@@ -0,0 +1,202 @@
+)
+
+var (
+	computeWASMInit = promauto.NewHistogramVec(prometheus.HistogramOpts{
+		Name: "compute_wasm_module_init",
+		Help: "how long it takes to initialize a WASM module",
+		Buckets: []float64{
+			float64(50 * time.Millisecond),
+			float64(100 * time.Millisecond),
+			float64(200 * time.Millisecond),
+			float64(500 * time.Millisecond),
+			float64(1 * time.Second),
+			float64(2 * time.Second),
+			float64(4 * time.Second),
+			float64(8 * time.Second),
+		},
+	}, []string{"workflowID", "stepRef"})
+	computeWASMExec = promauto.NewHistogramVec(prometheus.HistogramOpts{
+		Name: "compute_wasm_module_exec",
+		Help: "how long it takes to execute a request from a WASM module",
+		Buckets: []float64{
+			float64(50 * time.Millisecond),
+			float64(100 * time.Millisecond),
+			float64(200 * time.Millisecond),
+			float64(500 * time.Millisecond),
+			float64(1 * time.Second),
+			float64(2 * time.Second),
+			float64(4 * time.Second),
+			float64(8 * time.Second),
+		},
+	}, []string{"workflowID", "stepRef"})
+)
+
+var _ capabilities.ActionCapability = (*Compute)(nil)
+
+type Compute struct {
+	log      logger.Logger
+	registry coretypes.CapabilitiesRegistry
+	modules  *moduleCache
+}
+
+func (c *Compute) RegisterToWorkflow(ctx context.Context, request capabilities.RegisterToWorkflowRequest) error {
+	return nil
+}
+
+func (c *Compute) UnregisterFromWorkflow(ctx context.Context, request capabilities.UnregisterFromWorkflowRequest) error {
+	return nil
+}
+
+func generateID(binary []byte) string {
+	id := sha256.Sum256(binary)
+	return fmt.Sprintf("%x", id)
+}
+
+func (c *Compute) Execute(ctx context.Context, request capabilities.CapabilityRequest) (capabilities.CapabilityResponse, error) {
+	binary, err := c.popBytesValue(request.Config, binaryKey)
+	if err != nil {
+		return capabilities.CapabilityResponse{}, fmt.Errorf("invalid request: %w", err)
+	}
+
+	config, err := c.popBytesValue(request.Config, configKey)
+	if err != nil {
+		return capabilities.CapabilityResponse{}, fmt.Errorf("invalid request: %w", err)
+	}
+
+	id := generateID(binary)
+
+	m, ok := c.modules.get(id)
+	if !ok {
+		mod, err := c.initModule(id, binary, request.Metadata.WorkflowID, request.Metadata.ReferenceID)
+		if err != nil {
+			return capabilities.CapabilityResponse{}, err
+		}
+
+		m = mod
+	}
+
+	return c.executeWithModule(m.module, config, request)
+}
+
+func (c *Compute) initModule(id string, binary []byte, workflowID, referenceID string) (*module, error) {
+	initStart := time.Now()
+	mod, err := host.NewModule(&host.ModuleConfig{Logger: c.log}, binary)
+	if err != nil {
+		return nil, fmt.Errorf("failed to instantiate WASM module: %w", err)
+	}
+
+	mod.Start()
+
+	initDuration := time.Now().Sub(initStart)
+	computeWASMInit.WithLabelValues(workflowID, referenceID).Observe(float64(initDuration))
+
+	m := &module{module: mod}
+	c.modules.add(id, m)
+	return m, nil
+}
+
+func (c *Compute) popBytesValue(m *values.Map, key string) ([]byte, error) {
```

Comment:
nit: pop makes me think something is removed from values.  bytesValue would be fine.

Reply:
We do remove from it, see line 137 :)